ibmcloud: unit tests don't compile

While working on https://github.com/confidential-containers/cloud-api-adaptor/issues/47 I realized the tests `pkg/adaptor/hypervisor/ibmcloud/server_test.go` and `pkg/adaptor/hypervisor/ibmcloud/shim_test.go` don't compile.

The build errors:
```
[wmoschet@wainer-laptop ibmcloud]$ go test
# github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/ibmcloud [github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/ibmcloud.test]
./server_test.go:89:58: undefined: Server
./server_test.go:138:64: cannot use newAgentService() (type *agentService) as type grpc.AgentServiceService in argument to grpc.RegisterAgentServiceService:
	*agentService does not implement grpc.AgentServiceService (missing CreateSandbox method)
./server_test.go:159:58: undefined: Server
./server_test.go:163:19: too many arguments in call to NewServer
./server_test.go:163:47: undefined: ServiceConfig
./server_test.go:193:41: undefined: Server
./shim_test.go:444:63: undefined: grpc.StartTracingRequest
./shim_test.go:448:62: undefined: grpc.StopTracingRequest
./shim_test.go:456:59: undefined: grpc.CreateVMRequest
./shim_test.go:460:60: undefined: grpc.DestroyVMRequest
./server_test.go:163:19: too many errors
FAIL	github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/ibmcloud [build failed]
```

I did a couple of fixes as shown in the diff below:
```
diff --git a/pkg/adaptor/hypervisor/ibmcloud/server_test.go b/pkg/adaptor/hypervisor/ibmcloud/server_test.go
index c1ada6e..f98fd80 100644
--- a/pkg/adaptor/hypervisor/ibmcloud/server_test.go
+++ b/pkg/adaptor/hypervisor/ibmcloud/server_test.go
@@ -1,5 +1,6 @@
 // (C) Copyright IBM Corp. 2022.
 // SPDX-License-Identifier: Apache-2.0
+// +build ibmcloud
 
 package ibmcloud
 
@@ -21,6 +22,7 @@ import (
        "github.com/IBM/vpc-go-sdk/vpcv1"
 
        "github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/forwarder"
+       "github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor"
        daemon "github.com/confidential-containers/cloud-api-adaptor/pkg/forwarder"
        "github.com/confidential-containers/cloud-api-adaptor/pkg/podnetwork/tunneler"
        "github.com/confidential-containers/cloud-api-adaptor/pkg/util/http/upgrader"
@@ -86,7 +88,7 @@ func TestCreateStartAndStop(t *testing.T) {
        }
 }
 
-func testServerStart(t *testing.T, ctx context.Context) (Server, string, string, pb.HypervisorService, chan error) {
+func testServerStart(t *testing.T, ctx context.Context) (hypervisor.Server, string, string, pb.HypervisorService, chan error) {
        dir, err := ioutil.TempDir("", "helper")
        if err != nil {
                t.Fatal(err)
@@ -156,26 +158,41 @@ func startAgentServer(t *testing.T) string {
        return port
 }
 
-func newServer(t *testing.T, socketPath, podsDir string) Server {
+func newServer(t *testing.T, socketPath, podsDir string) hypervisor.Server {
+       hypervisorConfig := hypervisor.Config{
+               PodsDir:     podsDir,
+               SocketPath:  socketPath,
+               HypProvider: "ibmcloud",
+       }
        switch strings.ToLower(os.Getenv("USE_IBM_CLOUD")) {
        case "", "no", "false", "0":
                port := startAgentServer(t)
-               return NewServer(socketPath, &mockVpcV1{}, &ServiceConfig{}, &mockWorkerNode{}, podsDir, port)
+               return NewServer(hypervisorConfig, Config{}, &mockWorkerNode{}, port)
        }
        log.Print("Using IBM Cloud...")
        apiKey := os.Getenv("API_KEY")
        if apiKey == "" {
                t.Fatal("Specify the API key as API_KEY")
        }
-       vpcv1, err := NewVpcV1(apiKey)
-       if err != nil {
-               t.Fatal(err)
+
+       iamServiceURL := os.Getenv("IAM_SERVICE_URL")
+       if iamServiceURL == "" {
+               t.Fatal("Specify the IAM Service URL as IAM_SERVICE_URL")
        }
+
+       vpcServiceURL := os.Getenv("VPC_SERVICE_URL")
+       if vpcServiceURL == "" {
+               t.Fatal("Specify the VPC Service URL as VPC_SERVICE_URL")
+       }
+
        keyId := os.Getenv("KEY_ID")
        if keyId == "" {
                t.Fatal("Specify the SSH key ID as KEY_ID")
        }
-       serviceConfig := &ServiceConfig{
+
+       // TODO: need to remove the hard-coded config and finish the
+       // implementation in order to use vpcv1.
+       serviceConfig := Config{
                ProfileName:              "bx2-2x8",
                ZoneName:                 "us-south-2",
                ImageID:                  "r134-d2090805-5652-4845-b287-46232e1098c3",
@@ -185,12 +202,15 @@ func newServer(t *testing.T, socketPath, podsDir string) Server {
                SecondarySecurityGroupID: "r134-2d59206c-ac4d-4488-b9a4-a086bef59ee5",
                KeyID:                    keyId,
                VpcID:                    "r134-c199bf26-ec6d-4c5d-a0a2-1e74d312891f",
+               ApiKey:                   apiKey,
+               IamServiceURL:            iamServiceURL,
+               VpcServiceURL:            vpcServiceURL,
        }
 
-       return NewServer(socketPath, vpcv1, serviceConfig, &mockWorkerNode{}, podsDir, daemon.DefaultListenPort)
+       return NewServer(hypervisorConfig, serviceConfig, &mockWorkerNode{}, daemon.DefaultListenPort)
 }
 
-func testServerShutdown(t *testing.T, s Server, socketPath, dir string, serverErrCh chan error) {
+func testServerShutdown(t *testing.T, s hypervisor.Server, socketPath, dir string, serverErrCh chan error) {
 
        if err := s.Shutdown(); err != nil {
                t.Error(err)
diff --git a/pkg/adaptor/hypervisor/ibmcloud/shim_test.go b/pkg/adaptor/hypervisor/ibmcloud/shim_test.go
index d04c9fe..6bae051 100644
--- a/pkg/adaptor/hypervisor/ibmcloud/shim_test.go
+++ b/pkg/adaptor/hypervisor/ibmcloud/shim_test.go
@@ -1,5 +1,6 @@
 // (C) Copyright IBM Corp. 2022.
 // SPDX-License-Identifier: Apache-2.0
+// +build ibmcloud
 
 package ibmcloud
 
@@ -17,6 +18,7 @@ import (
        "testing"
        "time"
 
+       "github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor"
        daemon "github.com/confidential-containers/cloud-api-adaptor/pkg/forwarder"
        "github.com/confidential-containers/cloud-api-adaptor/pkg/podnetwork"
        "github.com/containerd/containerd/pkg/cri/annotations"
@@ -85,7 +87,13 @@ func TestShim(t *testing.T) {
                workerNode = podnetwork.NewWorkerNode(t, "")
        }
 
-       server := NewServer(helperSocketPath, &mockVpcV1{primaryIP: primaryIP, secondaryIP: secondaryIP}, &ServiceConfig{}, workerNode, podsDir, port)
+       hypervisorConfig := hypervisor.Config{
+               PodsDir:        podsDir,
+               SocketPath:     helperSocketPath,
+               HypProvider:    "ibmcloud",
+       }
+
+       server := NewServer(hypervisorConfig, &mockVpcV1{primaryIP: primaryIP, secondaryIP: secondaryIP}, &ServiceConfig{}, workerNode, port)
 
        serverDone := make(chan struct{})
        go func() {
```
... however after those changes there are still errors:

```
# github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/ibmcloud [github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/ibmcloud.test]
./server_test.go:140:64: cannot use newAgentService() (type *agentService) as type grpc.AgentServiceService in argument to grpc.RegisterAgentServiceService:
	*agentService does not implement grpc.AgentServiceService (missing CreateSandbox method)
./shim_test.go:96:21: too many arguments in call to NewServer
./shim_test.go:96:101: undefined: ServiceConfig
./shim_test.go:195:64: cannot use newAgentService() (type *agentService) as type grpc.AgentServiceService in argument to grpc.RegisterAgentServiceService:
	*agentService does not implement grpc.AgentServiceService (missing CreateSandbox method)
./shim_test.go:452:63: undefined: grpc.StartTracingRequest
./shim_test.go:456:62: undefined: grpc.StopTracingRequest
./shim_test.go:464:59: undefined: grpc.CreateVMRequest
./shim_test.go:468:60: undefined: grpc.DestroyVMRequest
FAIL	github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor/ibmcloud [build failed]

``` 
Unfortunately I cannot spend more time on this, so I am opening this issue.

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -49,9 +49,6 @@ jobs:
           cd cloud-api-adaptor
           make CLOUD_PROVIDER=${{ matrix.provider }} build
       - name: Test
-        # TODO: the tests for ibmcloud don't compile. Remove below guard when
-        # they get fixed (see https://github.com/confidential-containers/cloud-api-adaptor/issues/48).
-        if: ${{ matrix.provider != 'ibmcloud' }}
         working-directory: ${{ github.workspace }}/cloud-api-adaptor
         run: |
           go install github.com/jstemmer/go-junit-report@v1.0.0

--- a/pkg/adaptor/hypervisor/ibmcloud/server.go
+++ b/pkg/adaptor/hypervisor/ibmcloud/server.go
@@ -37,9 +37,15 @@ func NewServer(cfg hypervisor.Config, cloudCfg Config, workerNode podnetwork.Wor
 
 	logger.Printf("hypervisor config %v", cfg)
 	logger.Printf("cloud config %v", cloudCfg)
-	vpcV1, err := NewVpcV1(cloudCfg.ApiKey, cloudCfg.IamServiceURL, cloudCfg.VpcServiceURL)
-	if err != nil {
-		return nil
+
+	var vpcV1 VpcV1
+	if cloudCfg.ApiKey != "" {
+		//FIXME: Null ApiKey is used in unit tests
+		var err error
+		vpcV1, err = NewVpcV1(cloudCfg.ApiKey, cloudCfg.IamServiceURL, cloudCfg.VpcServiceURL)
+		if err != nil {
+			panic(err)
+		}
 	}
 
 	return &server{

--- a/pkg/adaptor/hypervisor/ibmcloud/server_test.go
+++ b/pkg/adaptor/hypervisor/ibmcloud/server_test.go
@@ -1,6 +1,7 @@
 // (C) Copyright IBM Corp. 2022.
 // SPDX-License-Identifier: Apache-2.0
 
+//go:build ibmcloud
 // +build ibmcloud
 
 package ibmcloud
@@ -23,6 +24,7 @@ import (
 	"github.com/IBM/vpc-go-sdk/vpcv1"
 
 	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/forwarder"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor"
 	daemon "github.com/confidential-containers/cloud-api-adaptor/pkg/forwarder"
 	"github.com/confidential-containers/cloud-api-adaptor/pkg/podnetwork/tunneler"
 	"github.com/confidential-containers/cloud-api-adaptor/pkg/util/http/upgrader"
@@ -88,7 +90,7 @@ func TestCreateStartAndStop(t *testing.T) {
 	}
 }
 
-func testServerStart(t *testing.T, ctx context.Context) (Server, string, string, pb.HypervisorService, chan error) {
+func testServerStart(t *testing.T, ctx context.Context) (hypervisor.Server, string, string, pb.HypervisorService, chan error) {
 	dir, err := ioutil.TempDir("", "helper")
 	if err != nil {
 		t.Fatal(err)
@@ -158,26 +160,33 @@ func startAgentServer(t *testing.T) string {
 	return port
 }
 
-func newServer(t *testing.T, socketPath, podsDir string) Server {
+func newServer(t *testing.T, socketPath, podsDir string) hypervisor.Server {
 	switch strings.ToLower(os.Getenv("USE_IBM_CLOUD")) {
 	case "", "no", "false", "0":
 		port := startAgentServer(t)
-		return NewServer(socketPath, &mockVpcV1{}, &ServiceConfig{}, &mockWorkerNode{}, podsDir, port)
+		cfg := hypervisor.Config{
+			SocketPath:  socketPath,
+			PodsDir:     podsDir,
+			HypProvider: "ibmcloud",
+		}
+		srv := NewServer(cfg, Config{}, &mockWorkerNode{}, port)
+		srv.(*server).service.(*hypervisorService).vpcV1 = &mockVpcV1{}
+		return srv
 	}
 	log.Print("Using IBM Cloud...")
 	apiKey := os.Getenv("API_KEY")
 	if apiKey == "" {
 		t.Fatal("Specify the API key as API_KEY")
 	}
-	vpcv1, err := NewVpcV1(apiKey)
-	if err != nil {
-		t.Fatal(err)
-	}
+
 	keyId := os.Getenv("KEY_ID")
 	if keyId == "" {
 		t.Fatal("Specify the SSH key ID as KEY_ID")
 	}
-	serviceConfig := &ServiceConfig{
+	serviceConfig := Config{
+		ApiKey:                   apiKey,
+		IamServiceURL:            "https://iam.cloud.ibm.com/identity/token",
+		VpcServiceURL:            "https://jp-tok.iaas.cloud.ibm.com/v1",
 		ProfileName:              "bx2-2x8",
 		ZoneName:                 "us-south-2",
 		ImageID:                  "r134-d2090805-5652-4845-b287-46232e1098c3",
@@ -189,10 +198,17 @@ func newServer(t *testing.T, socketPath, podsDir string) Server {
 		VpcID:                    "r134-c199bf26-ec6d-4c5d-a0a2-1e74d312891f",
 	}
 
-	return NewServer(socketPath, vpcv1, serviceConfig, &mockWorkerNode{}, podsDir, daemon.DefaultListenPort)
+	cfg := hypervisor.Config{
+		SocketPath:  socketPath,
+		PodsDir:     podsDir,
+		HypProvider: "ibmcloud",
+	}
+	srv := NewServer(cfg, serviceConfig, &mockWorkerNode{}, daemon.DefaultListenPort)
+
+	return srv
 }
 
-func testServerShutdown(t *testing.T, s Server, socketPath, dir string, serverErrCh chan error) {
+func testServerShutdown(t *testing.T, s hypervisor.Server, socketPath, dir string, serverErrCh chan error) {
 
 	if err := s.Shutdown(); err != nil {
 		t.Error(err)

--- a/pkg/adaptor/hypervisor/ibmcloud/shim_test.go
+++ b/pkg/adaptor/hypervisor/ibmcloud/shim_test.go
@@ -1,6 +1,7 @@
 // (C) Copyright IBM Corp. 2022.
 // SPDX-License-Identifier: Apache-2.0
 
+//go:build ibmcloud
 // +build ibmcloud
 
 package ibmcloud
@@ -19,6 +20,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor"
 	daemon "github.com/confidential-containers/cloud-api-adaptor/pkg/forwarder"
 	"github.com/confidential-containers/cloud-api-adaptor/pkg/podnetwork"
 	"github.com/containerd/containerd/pkg/cri/annotations"
@@ -87,18 +89,24 @@ func TestShim(t *testing.T) {
 		workerNode = podnetwork.NewWorkerNode(t, "")
 	}
 
-	server := NewServer(helperSocketPath, &mockVpcV1{primaryIP: primaryIP, secondaryIP: secondaryIP}, &ServiceConfig{}, workerNode, podsDir, port)
+	cfg := hypervisor.Config{
+		SocketPath:  helperSocketPath,
+		PodsDir:     podsDir,
+		HypProvider: "ibmcloud",
+	}
+	srv := NewServer(cfg, Config{}, workerNode, port)
+	srv.(*server).service.(*hypervisorService).vpcV1 = &mockVpcV1{primaryIP: primaryIP, secondaryIP: secondaryIP}
 
 	serverDone := make(chan struct{})
 	go func() {
 		defer close(serverDone)
 
-		if err := server.Start(ctx); err != nil {
+		if err := srv.Start(ctx); err != nil {
 			t.Error(err)
 		}
 	}()
 
-	<-server.Ready()
+	<-srv.Ready()
 
 	clientDone := make(chan error)
 
@@ -190,9 +198,6 @@ func startTestAgent(t *testing.T, ctx context.Context, agentSocketPath string, d
 	agent.RegisterHealthService(ttrpcServer, &healthService{})
 
 	socket := agentSocketPath
-	if len(socket) > 0 && socket[0] == '@' {
-		socket = socket + "\x00"
-	}
 
 	listener, err := net.Listen("unix", socket)
 	if err != nil {
@@ -443,6 +448,8 @@ func (s *agentService) AddARPNeighbors(ctx context.Context, req *agent.AddARPNei
 	log.Printf("agent call: %T %#v", req, req)
 	return &types.Empty{}, nil
 }
+
+/*
 func (s *agentService) StartTracing(ctx context.Context, req *agent.StartTracingRequest) (*types.Empty, error) {
 	log.Printf("agent call: %T %#v", req, req)
 	return &types.Empty{}, nil
@@ -451,15 +458,16 @@ func (s *agentService) StopTracing(ctx context.Context, req *agent.StopTracingRe
 	log.Printf("agent call: %T %#v", req, req)
 	return &types.Empty{}, nil
 }
+*/
 func (s *agentService) GetMetrics(ctx context.Context, req *agent.GetMetricsRequest) (*agent.Metrics, error) {
 	log.Printf("agent call: %T %#v", req, req)
 	return &agent.Metrics{}, nil
 }
-func (s *agentService) CreateVM(ctx context.Context, req *agent.CreateVMRequest) (*types.Empty, error) {
+func (s *agentService) CreateSandbox(ctx context.Context, req *agent.CreateSandboxRequest) (*types.Empty, error) {
 	log.Printf("agent call: %T %#v", req, req)
 	return &types.Empty{}, nil
 }
-func (s *agentService) DestroyVM(ctx context.Context, req *agent.DestroyVMRequest) (*types.Empty, error) {
+func (s *agentService) DestroySandbox(ctx context.Context, req *agent.DestroySandboxRequest) (*types.Empty, error) {
 	log.Printf("agent call: %T %#v", req, req)
 	return &types.Empty{}, nil
 }
@@ -496,7 +504,14 @@ func (s *agentService) AddSwap(ctx context.Context, req *agent.AddSwapRequest) (
 	log.Printf("agent call: %T %#v", req, req)
 	return &types.Empty{}, nil
 }
-
+func (s *agentService) GetVolumeStats(ctx context.Context, req *agent.VolumeStatsRequest) (*agent.VolumeStatsResponse, error) {
+	log.Printf("agent call: %T %#v", req, req)
+	return &agent.VolumeStatsResponse{}, nil
+}
+func (s *agentService) ResizeVolume(ctx context.Context, req *agent.ResizeVolumeRequest) (*types.Empty, error) {
+	log.Printf("agent call: %T %#v", req, req)
+	return &types.Empty{}, nil
+}
 func (s *agentService) PullImage(ctx context.Context, req *agent.PullImageRequest) (*types.Empty, error) {
 	log.Printf("agent call: PullImage %#v", req)
 	return &types.Empty{}, nil

